CAAM Black key support for asymmetric crypto

clementfaure opened the pull request:
# CAAM Black/Blob support in OPTEE

## CAAM Black key overview
The CAAM provides a key encapsulation mechanism called Black key. This is a protection of user keys against bus snooping attacks when the keys are being written or read from external memory (outside the SoC). The black key mechanism encapsulates/decapsulates keys on the fly in an encrypted data structure called a black key. Before copying the plain text key to memory, the CAAM automatically encrypts the key in the black key format using a key encryption key. This key encryption key is called JDKEK - Job ring descriptor key encryption key.

## CAAM Black key types
There are two different black key formats
 - AES-ECB : The key is encrypted using AES-ECB. The resulting black key will a multiple of 16 bytes. => AES-ECB(plain text key + zero padding)
 - AES-CCM : The key is encrypted using AES-CCM (slower than ECB). This format also includes a MAC to ensure integrity and authenticity. This black key is at least 12 bytes longer than the plain text key. Key modulo(8 bytes) + 6 bytes nonce + 6 bytes ICV => AES-CCM(plain text key + zero padding + nonce) + ICV

Please note that in the CAAM context, a plain text key is called a Red key.
The CAAM driver requires the user to specify the given key type in the job descriptor. The CAAM cannot tell if the key given is Red or Black-ECB/CCM.
Once encapsulated, a black key cannot be decapsulated to memory. The CAAM will prevent any Black key to plain text key in memory operation. A black key can only be decapsulated to CAAM key registers.

## JDKEK
Each job ring has its own key encryption key to encapsulate keys into black key format. This key is randomly re-generated upon each reset. The black key encapsulation is modified via the appropriate Trustzone status and SDID value. With this, each domain has its own black keys and cannot use black keys generated by other domains. When the CAAM is instructed to use a black key, the CAAM automatically decrypts the incoming black key in its own key register to do the crypto operations.
JDKEK has a limited lifetime, and it is re-generated (randomly) upon each power cycle (reset, suspend/resume). A black key generated in a power cycle will be unusable after the next power cycle.
Please note that trusted descriptor key encryption keys - TDKEK can also be used to encapsulate black keys, but this is not supported in this PR.

## Black blobs
To retain encrypted keys across power cycles, the CAAM is able to encrypt keys using a secure non-volatile key encryption key. This non-volatile KEK is device specific and cannot be extracted from the device. The non-volatile KEK is derived from the device master key. There are several methods to define the master key.
Black blobs provide confidentiality and integrity protection, and keys can be stored in a long-term storage. To make a black key power cycle proof, the CAAM generates a random blob key to encrypt the black key. This random blob key is then encrypted with the non-volatile Blob KEK.

A black key can be encapsulated into a black blob, and a black blob can be decapsulated to a black key (in memory) or plain text key (CAAM key registers).
```

                                                                BKEK
                                                                  │
                                                                  ▼
                                                            ┌────────────┐
                                                            │ AES-ECB    │
                     JDKEK or               RNG ───────────►│ Encryption │
                       TDKEK                 │              └─────┬──────┘
                         │                   │ Blob key           │
                         │                   │                    ▼
                         │                   │              ┌────────────┐
                         ▼                   ▼              │  Key blob  │
                  ┌─────────────┐      ┌────────────┐       ├────────────┤
┌───────────┐     │ AES-ECB/CCM │      │ AES-CCM    │       │            │
│ Black key ├────►│  Decryption ├─────►│ Encryption ├──────►│ Data blob  │
└───────────┘     └─────────────┘      └──────────┬─┘       │            │
                                                  │         │            │
                                                  │         ├────────────┤
                                                  └────────►│   MAC      │
                                                            └────────────┘
```

# CAAM Black key/blob in OPTEE

## Pull-request overview
This PR aims at adding the support of Black key/blob to CAAM asymmetric cryptographic operations. The key generation and the cryptographic operations being linked by the same API vectors, the support for asymmetric crypto does not require modification in the OPTEE core.
This PR also provides to the CAAM driver the ability to handle plain text keys and black keys. Indeed, the CAAM driver is able to use a previously generated black key or a plain text key given by the user as an input.
Please note that, in the case of asymmetric keys, only the private part of the key pair is encapsulated as a black key/blob.

## Black key supported formats
This PR adds the support for AES-ECB and AES-CCM black keys. Only one key generation format is supported at the same time. It must be chosen at compilation. It defines the default format for CAAM key generation.
There are 3 formats available:
 - Plain text
 - AES-ECB
 - AES-CCM
Even if the CAAM supports only one key generation format by default, the CAAM is still able to use any key format (plain text, AES-ECB/CCM ...) for its crypto operations.

I would advise sticking to AES-CCM. It provides better security, adds authenticity and performance regarding the black key encapsulation/decapsulation are similar to AES-ECB.

Because a cryptographic key pair is a TEE object and can be saved in persistent memory at any time, private key pairs will always be stored as a black blob (which is power cycle proof).
```
┌─────────────────────────┐
│                         │
│   CAAM Key generation   ├──────────────►   Key format defined by `caam_key_default_key_gen_type()`
│                         │
└─────────────────────────┘

┌─────────────────────────┐
│                         │
│  CAAM Crypto operations │◄──────────────   Plain text key, black key ECB, CCM, ...
│                         │
└─────────────────────────┘
```

## How to handle both Black keys and plain text keys?
Along the key data, a header with some metadata must be added to tell the CAAM how to handle the incoming key.
Metadata:
 - Magic number (32 bits): a magic number the CAAM driver can detect on the first 32 bits of the key buffer is needed. If a magic number is detected, the key is a black key generated by the CAAM. If not, this is a plain text key.
 - Key type (32 bits): specify the CAAM black key type. This info is needed by the CAAM for the black key decapsulation.
 - Key security size (32 bits): this is the security key size of the black key. This is different from the actual buffer key size. This info is needed by the CAAM for the black key decapsulation.

When generating a key pair, the CAAM driver encapsulates the generated private key in the following format `struct caam_key_serialized {}` :

```
Black blob key generated by the CAAM:
+-----------------------------------------------------+
| MAGIC number - 0xCAAFBFFB (32 bits)                 |
+-----------------------------------------------------+
| Key type (if magic number is present)               |
| -  CAAM_KEY_PLAIN_TEXT                              |
| -  CAAM_KEY_BLACK_ECB                               |
| -  CAAM_KEY_BLACK_CCM                               |
+-----------------------------------------------------+
| Plain text key size (if magic number is present)    |
+-----------------------------------------------------+
| Black blob                                          |
+-----------------------------------------------------+
```

```
Regular key generated by the CAAM:
+-----------------------------------------------------+
| Plain text data                                     |
+-----------------------------------------------------+
```

## Black blobs as a bignum in OPTEE core
Because a black key or blob is always bigger than the original plain text key size, the maximum bignum size must be increased. This is especially true for RSA-4096 key pair. `CFG_CORE_BIGNUM_MAX_BITS` value must be increased accordingly.

The private part of the key pair can be any size : the OPTEE core does not rely on the buffer size of the private key to determine the key pair size. The CAAM black key size overhead is not a problem for asymmetric crypto.

## Supported algos?
This PR adds the black key/blob support for the following algorithms:
 - DH
 - DSA
 - RSA
 - ECC
The CAAM directly outputs the private part of the key pair encapsulated in a black key format. The plain text private key is never exposed in the memory.
Note for RSA: currently, the RSA key pair generation CAAM function outputs the private part of the key pair in plain text format to memory. The plain text private key is manually encapsulated as a black key right after key pair generation. This will be solved in the future with the support of CAAM internal secure memory.
For the other algorithms, the key generation operations directly output the private key part in a black format.

## What about the support for symmetric algos?
The black key support for symmetric algos is another story and will most likely require changes in OPTEE Core. There are multiples obstacles:

 1. **The key buffer size is not equal to the key security size.**
 In the case of an encapsulated symmetric key, the actual key buffer size does not reflect the actual security key size. For the CAAM, a 256 bits key requires at least a 640 bits buffer to encapsulate the key as a black blob.
 An HW crypto driver could also refer to key IDs if it has its own keystore/internal memory (like the SE050).

 One way to work around this core/GP limitation would be to have a crypto HW driver callback that returns the actual security key size of the given encapsulated key. This way, the OPTEE core would ask the HW crypto driver for its key size checks ... If such callback is not implemented by the HW driver, the core would rely on the buffer size to determine the security key size.

 2.**The black key generation cannot be done through TEE Crypto API**
 There is no clean way to specify the crypto HW driver the type of key to generate. Going through `TEE_GenerateKey()` would require to add new `TEE_TYPE_...` like `TEE_TYPE_AES_BLACK_KEY` or something ... which is not very GP compliant.
 The only way to generate black keys would be a PTA.

Feel free to ask questions.

Regards,
Clement

jenswi-linaro commented:
What are black keys?

sahilnxp commented:
> What are black keys?

@jenswi-linaro That is CAAM terminology for Runtime Secure Keys.

jenswi-linaro commented:
Yes, but how are they used? I need to understand what I'm reviewing. I'd rather not read a spec of 1000 pages first.

sahilnxp commented:
Sorry for that, It is in context of `[RFC] Runtime Secure Keys Framework Proposal` https://github.com/OP-TEE/optee_os/pull/4807

clementfaure commented:
Hi @jenswi-linaro , I will provide a description of the CAAM black key mechanism shortly

clementfaure commented:
Hello all, please have a look at PR description for more info about the black implementation.

jenswi-linaro reviewed (commented):
Comments for "drivers: caam: add CAAM key driver"

clementfaure commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,739 @@
+	assert(data && size);
+	assert(size > sizeof(key.header));
+
+	memcpy(&key, data, sizeof(key.header));
```
This corresponds to an older design, I will fix it indeed.

clementfaure commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,739 @@
+	assert(data && size);
+	assert(size > sizeof(key.header));
+
+	if (get_magic_number(data, size) != MAGIC_NUMBER)
```
A plain text key, in this design, doesn't feature any header. It's the magic number that defines if a key is plain or encrypted.

clementfaure replied:
After adding `data_to_serialized_key()`, I must keep it that way

clementfaure commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,739 @@
+	case CAAM_KEY_BLACK_ECB:
+		/* ECB-black key must be a multiple of 16 bytes */
+		return ROUNDUP(key->sec_size, 16);
+		break;
```
Ok

clementfaure commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,739 @@
+	}
+}
+
+void caam_key_to_str(const char *trace , const struct caamkey *key)
```
Ok

clementfaure commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,739 @@
+		return;
+
+	KEY_TRACE("%s key_type:%s key_size:%zu is_blob:%s addr:%p",
+		  caam_key_type_to_str[key->key_type], trace, key->sec_size,
```
Ok, I will do it

clementfaure commented:
I will wait for additional feedback/review before addressing comments. Same thing for checkpatch issues. By the way, I'm not sure about this check:
```
CHECK: Macro argument reuse 'alg' - possible side-effects?
#2702: FILE: core/drivers/crypto/caam/include/caam_desc_helper.h:515:
+#define SHARED_SECRET(type, alg)                                               \
 	(CMD_OP_TYPE | OP_TYPE(UNI) | PROTID(SHARED_SECRET) |                  \
+	 PROT_PK_TYPE(type) | PROT_PRI_ENC(alg) | PROT_PRI_EXT(alg))
```
Is it really a problem?

jforissier commented:
> I will wait for additional feedback/review before addressing comments. Same thing for checkpatch issues. By the way, I'm not sure about this check:
> 
> ```
> CHECK: Macro argument reuse 'alg' - possible side-effects?
> #2702: FILE: core/drivers/crypto/caam/include/caam_desc_helper.h:515:
> +#define SHARED_SECRET(type, alg)                                               \
>  	(CMD_OP_TYPE | OP_TYPE(UNI) | PROTID(SHARED_SECRET) |                  \
> +	 PROT_PK_TYPE(type) | PROT_PRI_ENC(alg) | PROT_PRI_EXT(alg))
> ```
> 
> Is it really a problem?

In this case I think you could easily use a temporary and the `({ .. })` syntax like here:

https://github.com/OP-TEE/optee_os/blob/23419643ecf2dcce31ccc2ee03dfa74096a45de5/core/include/signed_hdr.h#L183-L184

clementfaure commented:
> > I will wait for additional feedback/review before addressing comments. Same thing for checkpatch issues. By the way, I'm not sure about this check:
> > ```
> > CHECK: Macro argument reuse 'alg' - possible side-effects?
> > #2702: FILE: core/drivers/crypto/caam/include/caam_desc_helper.h:515:
> > +#define SHARED_SECRET(type, alg)                                               \
> >  	(CMD_OP_TYPE | OP_TYPE(UNI) | PROTID(SHARED_SECRET) |                  \
> > +	 PROT_PK_TYPE(type) | PROT_PRI_ENC(alg) | PROT_PRI_EXT(alg))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Is it really a problem?
> 
> In this case I think you could easily use a temporary and the `({ .. })` syntax like here:
> 
> https://github.com/OP-TEE/optee_os/blob/23419643ecf2dcce31ccc2ee03dfa74096a45de5/core/include/signed_hdr.h#L183-L184

Unfortunately, it doesn't work
```

In file included from core/drivers/crypto/caam/acipher/../include/caam_common.h:11,
                 from core/drivers/crypto/caam/acipher/../include/caam_acipher.h:10,
                 from core/drivers/crypto/caam/acipher/caam_rsa.c:8:
core/drivers/crypto/caam/acipher/caam_rsa.c: In function ‘do_gen_keypair’:
core/drivers/crypto/caam/acipher/caam_rsa.c:668:69: error: ‘NONE’ undeclared (first use in this function)
  668 |                         caam_desc_add_word(desc, RSA_FINAL_KEY(ALL, NONE));
      |                                                                     ^~~~
core/drivers/crypto/caam/acipher/../include/caam_desc_helper.h:488:19: note: in definition of macro ‘RSA_FINAL_KEY’
  488 |         ({ typeof(alg) _alg = (alg);                                           \
      |                   ^~~
```
`alg` is not a variable but a macro extension `#define PROT_PRI_ENC(alg)	SHIFT_U32(PROT_PRI_ENC_##alg, 2)`

clementfaure commented:
Force-push:
* Address some comments. Some comments are still open to discussion.
* Rebase

jenswi-linaro commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,739 @@
+		uint32_t sec_size;     /* The original plain text key size */
+	} header;
+	uint8_t key[];
+} __packed;
```
The proper way to do that is to make a struct layout that is free from padding by nature and then add a `static_assert()` to make sure that there really isn't any unexpected padding. Dropping the inner struct should be sufficient.

That leads to more efficient machine code, better warnings, and also saves some time for maintainers since there's less chance that others copy the slightly flawed `__packed` pattern. :-)

clementfaure replied:
If I drop `__packed` I will get a cast warning
```

core/drivers/crypto/caam/caam_key.c: In function ‘data_to_serialized_key’:
core/drivers/crypto/caam/caam_key.c:99:15: warning: cast increases required alignment of target type [-Wcast-align]
   99 |         key = (struct caam_key_serialized *)data;
      |               ^
core/drivers/crypto/caam/caam_key.c: At top level:
cc1: note: unrecognized command-line option ‘-Wno-c2x-extensions’ may have been intended to silence earlier diagnostics
```
A static assert can make sure it won't happen, but I will still get the warning, right?

jenswi-linaro replied:
You can cast it to `void *` instead since you have checked that the address is well aligned.

clementfaure replied:
I would personally go with either `__packed` or a `memcpy()`. I find it tricky to make a `void *` cast dependent of `assert()` in another line of code somewhere else.
I guess if we go that way, it must be properly documented in the code.

jenswi-linaro replied:
Feel free to add a comment.

clementfaure replied:
On?

jenswi-linaro replied:
> I guess if we go that way, it must be properly documented in the code.

clementfaure replied:
https://github.com/OP-TEE/optee_os/pull/5990/files#diff-2af92fa52b0087bf848ecd8b9b9de69c7330aa1b63cb31c3388387e1dcdb5bd7R109

jenswi-linaro replied:
I'm sorry, I don't understand, what are you proposing?

clementfaure replied:
Oops sorry, wrong link. I mean, I went with the `(void *)` cast and add comments : https://github.com/OP-TEE/optee_os/blob/21b78605d29a8a1d62f4cc483607ae429ccf7e94/core/drivers/crypto/caam/caam_key.c#L109

jenswi-linaro commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,721 @@
+	assert(size > sizeof(key->header));
+	assert(IS_ALIGNED_WITH_TYPE(data, struct caam_key_serialized));
+
+	key = (struct caam_key_serialized *)data;
```
`return (struct caam_key_serialized *)data;` ?

clementfaure replied:
Ok

jenswi-linaro replied:
OK what?

jenswi-linaro commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,721 @@
+	if (key->key_type != CAAM_KEY_PLAIN_TEXT) {
+		size_t alloc = 0;
+		struct caam_key_serialized key_serialized = { };
+
```
Empty line isn't needed.

clementfaure replied:
ok

jenswi-linaro commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,739 @@
+	assert(data && size);
+	assert(size > sizeof(key.header));
+
+	if (get_magic_number(data, size) != MAGIC_NUMBER)
```
Who generates the key? If it's only done by the CAAM then we could check for that and regenerate in the unlikely event. If it can be supplied externally then we have to assume that it might be crafted.

jenswi-linaro replied:
If you allow externally supplied plain text private keys then you'd need a magic larger than a plain text key to be able to block forged keys.

As I understand it, it's not about probability like hash collisions, it seems trivial to forge a plain text private key with the magic number as a prefix.

clementfaure replied:
> If you allow externally supplied plain text private keys then you'd need a magic larger than a plain text key to be able to block forged keys.

Yes, I want to maintain support for external plain text keys
> 
> As I understand it, it's not about probability like hash collisions, it seems trivial to forge a plain text private key with the magic number as a prefix.

Forging and trying to make a plain text key look like a black key, won't do much. If the key is forged to look like an ECB Black key, the singing operation will output a corrupted result. If the key is forged to look like a CCM Black key, the import key will fail (because the MAC verification) and no signing operation will be done.

jenswi-linaro replied:
OK, so it shouldn't be a security problem then.

jenswi-linaro commented:
> `alg` is not a variable but a macro extension `#define PROT_PRI_ENC(alg) SHIFT_U32(PROT_PRI_ENC_##alg, 2)`

The drawback with this approach is that cscope and grep become a bit less efficient. How about spelling out the constants instead?

clementfaure commented:
> PROT_PRI_ENC

I agree 

> > `alg` is not a variable but a macro extension `#define PROT_PRI_ENC(alg) SHIFT_U32(PROT_PRI_ENC_##alg, 2)`
> 
> The drawback with this approach is that cscope and grep become a bit less efficient. How about spelling out the constants instead?

The CAAM already has an extensive use of macro extension when it comes to build job descriptor, so I didn't think it was a problem. What do you suggest? I would like to avoid adding `if else` conditions ..

jenswi-linaro commented:
> > > `alg` is not a variable but a macro extension `#define PROT_PRI_ENC(alg) SHIFT_U32(PROT_PRI_ENC_##alg, 2)`
> > 
> > 
> > The drawback with this approach is that cscope and grep become a bit less efficient. How about spelling out the constants instead?
> 
> The CAAM already has an extensive use of macro extension when it comes to build job descriptor, so I didn't think it was a problem. What do you suggest? I would like to avoid adding `if else` conditions ..

In this case, it's clearly a false positive so I suppose we could ignore it for now. Perhaps someone can find a way later to avoid such false positives.

clementfaure commented:
Up

clementfaure commented:
Could you please remove the stall label? Thanks

clementfaure commented:
@embetrix Enable the CAAM driver and all private keys generated by the CAAM will be encrypted.

clementfaure commented:
@embetrix Have you tested on your side? Any feedback?

clementfaure commented:
@jforissier @jenswi-linaro I know that's a big chunk of code to review, but if you could spend some time on it, that would be great. Feel free to ask questions if this PR remains unclear to you. Thanks a lot.

jforissier commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,721 @@
+
+/*
+ * CAAM Key magic number.
+ * This 32 bits value added to the first 32 bits tells the CAAM driver if
```
"Added" sounds a bit strange here, how about "When the first 32 bits of a key buffer are equal to this value, the buffer is a serialized CAAM key structure"?

More fundamentally, what happens if a "plain" key happens to have MAGIC_NUMBER in its first 32-bits?

clementfaure replied:
I will rephrase it. Regarding your question, let me know if this previous discussion answers your question : https://github.com/OP-TEE/optee_os/pull/5990#discussion_r1189978494

jforissier replied:
> I will rephrase it. Regarding your question, let me know if this previous discussion answers your question : [#5990 (comment)](https://github.com/OP-TEE/optee_os/pull/5990#discussion_r1189978494)

It does, thanks ;) would you mind adding a comment to explain that collisions are not expected to be a problem, both for plain key generation, and regarding potential attacks (you comment [here](https://github.com/OP-TEE/optee_os/pull/5990#discussion_r1219291868) is especially useful IMO).

clementfaure replied:
Done

jforissier commented on core/drivers/crypto/caam/crypto.mk:
```diff
@@ -132,6 +132,12 @@ CFG_NXP_CAAM_RSA_KEY_FORMAT ?= 3
+# maximum size needed.
+# CAAM_KEY_PLAIN_TEXT    -> 4096 bits
+# CAAM_KEY_BLACK_ECB|CCM -> 4156 bits
+CFG_CORE_BIGNUM_MAX_BITS = 4156
```
`?=`

Are 4096 and 4156 the only valid values? If so it may be good to check it with `ifneq($(CFG_CORE_BIGNUM_MAX_BITS),...) ... $(error ...) ... endif`

clementfaure replied:
4096 and 4156 are the minimum values required for 4096 RSA keypairs and black keypairs. You could still be able to 
decrease these values if you don't want to generate 4096 keypairs.

Here I check if `CFG_CORE_BIGNUM_MAX_BITS` can store a 4096 bits RSA black keypair
https://github.com/OP-TEE/optee_os/pull/5990/files#diff-2af92fa52b0087bf848ecd8b9b9de69c7330aa1b63cb31c3388387e1dcdb5bd7R723-R727

clementfaure replied:
4096 and 4156 are the minumum value s

Here I check if `CFG_CORE_BIGNUM_MAX_BITS` can store a 4096 bits RSA black keypair
https://github.com/OP-TEE/optee_os/pull/5990/files#diff-2af92fa52b0087bf848ecd8b9b9de69c7330aa1b63cb31c3388387e1dcdb5bd7R723-R727

clementfaure replied:
I have to set it like this `=` instead of `?=` otherwise it won't override the default value.

jforissier replied:
But that's not allowed (see [here](https://optee.readthedocs.io/en/latest/building/gits/optee_os.html#there-are-only-two-ways) and a in the paragraphs above for details). It's either `?=` or `$(call force,CFG_CORE_BIGNUM_MAX_BITS,4156)`. In this case, if a default value is set somewhere else, you should try to set it earlier so that it becomes the actual default.

clementfaure replied:
Fix suggestion here : ba8ef3776

jforissier commented on core/drivers/crypto/caam/include/caam_desc_helper.h:
```diff
@@ -387,6 +387,13 @@ static inline void dump_desc(uint32_t *desc)
+ */
+#define LD_KEY(cla, dst, len)                                                  \
+	(CMD_KEY_TYPE | CMD_CLASS(cla) | KEY_DEST(dst) |                       \
+	 KEY_LENGTH(len))
```
Fits on the previous line I think

jforissier commented on core/drivers/crypto/caam/include/caam_key.h:
```diff
@@ -0,0 +1,155 @@
+ * Returns the default key type for CAAM key generation.
+ * The CAAM can only generate one key type.
+ */
+inline enum caam_key_type caam_key_default_key_gen_type(void)
```
`static inline`

clementfaure replied:
Yup

jforissier commented on core/drivers/crypto/caam/include/caam_trace.h:
```diff
@@ -279,6 +280,29 @@
+	do {                                                                   \
+		KEY_TRACE("KEY Descriptor");                                   \
+		DRV_DUMPDESC(desc);                                            \
+	} while (0)
```
Preferably move the line continuation symbols `\` to the left (align on the next tab stop after the "KEY Descriptor" line)

clementfaure replied:
I wanted to align with the style in the file but okay, I will fix

jforissier reviewed (commented):
A few comments on "drivers: caam: add CAAM key driver"

jforissier commented:
For commits:
* drivers: caam: add class field to FIFO_ST macro
* drivers: caam: add missing header
* drivers: caam: fix DSA_DUMPDESC macro

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier commented on core/drivers/crypto/caam/acipher/caam_rsa.c:
```diff
@@ -539,18 +546,32 @@ static TEE_Result do_gen_keypair(struct rsa_keypair *key, size_t key_size)
+
+	/* First allocate primes p and q */
+	genkey.p.key_type = CAAM_KEY_PLAIN_TEXT;
+	genkey.p.sec_size = (key_size / 8) / 2;
```
It might be good to have a macro for that, so the `/ 2` could be documented.

clementfaure replied:
p, q, dp, dq, qp are half the size of the keypair size (defined by n, d). I will define a macro if you want.

jforissier reviewed (commented):
Commenting on "drivers: caam: add CAAM key support for RSA"

clementfaure commented on core/drivers/crypto/caam/acipher/caam_prime_rsa.c:
```diff
@@ -373,6 +373,9 @@ static void do_desc_setup(uint32_t *desc, struct prime_data_rsa *data,
 			  const struct caambuf *small_prime,
 			  const paddr_t desc_prime)
 {
+	size_t key_size_bits = data->key_size / 2;
```
That's my mistake. It does not make sense to divide it by 2 here. I will divide it by 2 here instead: 
`caam_desc_add_word(desc, 5 * data->key_size/ 2);`

clementfaure replied:
It does not make sense to divide the key size here indeed. I will divide it in the function input argument.

clementfaure commented:
Force-push: rebase and tags applied

clementfaure commented:
Hi, can we re-open the PR?

jforissier commented:
> Hi, can we re-open the PR?

Sure.

clementfaure commented:
Any more comments on this pull-request?

jenswi-linaro commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,746 @@
+		if (!alloc)
+			return CAAM_FAILURE;
+
+		*size = alloc + sizeof(key_serialized);
```
`sizeof(struct caam_key_serialized)`, and remove the otherwise unused `key_serialized`.

clementfaure replied:
Indeed

jenswi-linaro commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,746 @@
+	if (key_type == CAAM_KEY_BLACK_ECB)
+		caam_desc_add_word(desc, FIFO_ST(CLASS_NO, PKHA_E_AES_ECB_JKEK,
+						 key->sec_size));
+	else if (key_type == CAAM_KEY_BLACK_CCM)
```
Add nothing if `key_type` isn't `CAAM_KEY_BLACK_CCM` either?

clementfaure replied:
If the key isn't black, no encapsulation is needed. There is an `assert()` at the beginning of the `caam_key_black_encapsulation()` function.

jenswi-linaro replied:
Please skip the "if (key_type == CAAM_KEY_BLACK_CCM)" part and only use an `else`.

If the answer is the same for all the other similar comments I've made, do the same there.

clementfaure replied:
Ok for this one.

jenswi-linaro replied:
When you're testing for as many different values as in the last case I'd prefer a switch-case construct, with a default for error handling.

At this location, I'd prefer just
``` C
if (key_type == CAAM_KEY_BLACK_ECB)
         ...
else
         ...
```
Please try to avoid testing for the same thing twice in the same function since it then in the second test "can't happen" if it was an error condition.

clementfaure replied:
Ok, I will check other places where `key_type` is tested

jenswi-linaro reviewed (commented):
Comments for "drivers: caam: add CAAM key driver"

There are a few checkpatch complaints that aren't false positives.

jenswi-linaro commented on core/drivers/crypto/caam/acipher/caam_rsa.c:
```diff
@@ -618,29 +649,39 @@ static TEE_Result do_gen_keypair(struct rsa_keypair *key, size_t key_size)
+			caam_desc_add_word(desc, RSA_FINAL_KEY(ALL, NONE));
+		else if (key_type == CAAM_KEY_BLACK_ECB)
+			caam_desc_add_word(desc, RSA_FINAL_KEY(ALL, ECB));
+		else if (key_type == CAAM_KEY_BLACK_CCM)
```
Add no word, if `key_type` isn't `CAAM_KEY_BLACK_CCM` either?

Same below.

clementfaure replied:
`key_type` have three valid values. This `if .. else if` covers all valid values.

jenswi-linaro commented on core/drivers/crypto/caam/acipher/caam_rsa.c:
```diff
@@ -105,17 +112,13 @@ static void do_keypair_free(struct caam_rsa_keypair *key)
-	}
+	caam_key_free(&key->d);
+	caam_key_free(&key->p);
+	caam_key_free(&key->q);
```
Did we leak memory here before?

jenswi-linaro commented on core/drivers/crypto/caam/acipher/caam_rsa.c:
```diff
@@ -703,19 +757,19 @@ static TEE_Result do_oaep_decoding(struct drvcrypt_rsa_ed *rsa_data)
-	struct caambuf maskedDB = { };
-	struct caambuf maskedSeed = { };
-	struct caambuf EM = { };
+	struct caambuf DB = {};
```
Why are you changing this?

jenswi-linaro commented on core/drivers/crypto/caam/acipher/caam_rsa.c:
```diff
@@ -1413,7 +1512,24 @@ static TEE_Result do_caam_decrypt(struct drvcrypt_rsa_ed *rsa_data,
+	else if (g_key_type == CAAM_KEY_BLACK_CCM)
+		operation |= PROT_RSA_KEY_ENC(CCM);
+	else if (g_key_type == CAAM_KEY_PLAIN_TEXT)
+		operation |= PROT_RSA_KEY_ENC(NONE);
```
Do nothing if `g_key_type` isn't `CAAM_KEY_PLAIN_TEXT` either?

clementfaure replied:
I make sure `g_key_type` has a valid value above.

clementfaure replied:
Sorry, yes, I will remove
```
else if (g_key_type == CAAM_KEY_PLAIN_TEXT)
		operation |= PROT_RSA_KEY_ENC(NONE);
```

jenswi-linaro reviewed (commented):
Comments for "drivers: caam: add CAAM key support for RSA"

jenswi-linaro commented on core/drivers/crypto/caam/acipher/caam_dsa.c:
```diff
@@ -430,9 +435,15 @@ static TEE_Result do_gen_keypair(struct dsa_keypair *key, size_t l_bits,
+		caam_desc_add_word(desc, PK_KEYPAIR_GEN(DL, NONE));
+	else if (key_type == CAAM_KEY_BLACK_ECB)
+		caam_desc_add_word(desc, PK_KEYPAIR_GEN(DL, ECB));
+	else if (key_type == CAAM_KEY_BLACK_CCM)
```
What if `key_type` isn't `CAAM_KEY_BLACK_CCM` either?

clementfaure replied:
Same, I check if `key_type` has a valid value.

jenswi-linaro reviewed (commented):
Comment for "drivers: caam: add CAAM key support for DSA"

jenswi-linaro commented on core/drivers/crypto/caam/acipher/caam_ecc.c:
```diff
@@ -258,45 +263,61 @@ static TEE_Result do_gen_keypair(struct ecc_keypair *key, size_t key_size)
+		caam_desc_add_word(desc, PK_KEYPAIR_GEN(ECC, NONE));
+	else if (key_type == CAAM_KEY_BLACK_ECB)
+		caam_desc_add_word(desc, PK_KEYPAIR_GEN(ECC, ECB));
+	else if (key_type == CAAM_KEY_BLACK_CCM)
```
What if `key_type` isn't `CAAM_KEY_BLACK_CCM` either?

jenswi-linaro commented on core/drivers/crypto/caam/acipher/caam_ecc.c:
```diff
@@ -660,11 +694,17 @@ static TEE_Result do_shared_secret(struct drvcrypt_secret_data *sdata)
+		caam_desc_add_word(desc, SHARED_SECRET(ECC, NONE));
+	else if (ecckey.d.key_type == CAAM_KEY_BLACK_ECB)
+		caam_desc_add_word(desc, SHARED_SECRET(ECC, ECB));
+	else if (ecckey.d.key_type == CAAM_KEY_BLACK_CCM)
```
What if `ecckey.d.key_type` isn't `CAAM_KEY_BLACK_CCM` either.

jenswi-linaro reviewed (commented):
Comments for "drivers: caam: add CAAM key support for ECC"

jenswi-linaro commented on core/drivers/crypto/caam/acipher/caam_dh.c:
```diff
@@ -226,6 +226,14 @@ static TEE_Result do_gen_keypair(struct dh_keypair *key,
+	 */
+	enum caam_key_type key_type = CAAM_KEY_PLAIN_TEXT;
+
+	if (key_type == CAAM_KEY_MAX_VALUE)
```
Why?

clementfaure replied:
Ok so, currently xtest `xtest_tee_test_4007_dh()` does not support CAAM black keys. During DH keypair generation, xtest checks if the returned buffer size is equal to the DH key size. A CAAM black key being bigger than the key itself, one of the check fails.
For xtest compatibility, DH generation is forced to plain text.
Let me know what you think

jenswi-linaro replied:
It seems that the test is broken. Is it possible to fix the test? It would be a shame to enforce limits here due to broken tests.

jenswi-linaro commented on core/drivers/crypto/caam/acipher/caam_dh.c:
```diff
@@ -301,9 +313,15 @@ static TEE_Result do_gen_keypair(struct dh_keypair *key,
+		caam_desc_add_word(desc, PK_KEYPAIR_GEN(DL, NONE));
+	else if (key_type == CAAM_KEY_BLACK_ECB)
+		caam_desc_add_word(desc, PK_KEYPAIR_GEN(DL, ECB));
+	else if (key_type == CAAM_KEY_BLACK_CCM)
```
What if `key_type` isn't `CAAM_KEY_BLACK_CCM` either?

jenswi-linaro commented on core/drivers/crypto/caam/acipher/caam_dh.c:
```diff
@@ -432,19 +447,25 @@ static TEE_Result do_shared_secret(struct drvcrypt_secret_data *sdata)
+		caam_desc_add_word(desc, SHARED_SECRET(DL, NONE));
+	else if (caam_dh_key.x.key_type == CAAM_KEY_BLACK_ECB)
+		caam_desc_add_word(desc, SHARED_SECRET(DL, ECB));
+	else if (caam_dh_key.x.key_type == CAAM_KEY_BLACK_CCM)
```
What if `caam_dh_key.x` isn't `CAAM_KEY_BLACK_CCM` either?

jenswi-linaro reviewed (commented):
Comments for "drivers: caam: add CAAM key support for DH"

jenswi-linaro reviewed (commented):
For commit "core: move CFG_CORE_BIGNUM_MAX_BITS default definition"
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

This commit should probably be first in the pull request.

clementfaure commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,746 @@
+	if (!key)
+		return;
+
+	if (key->buf.nocache == 0)
```
Ok

clementfaure commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,746 @@
+	 * If a security key size is given, use it. Otherwise, rely on
+	 * the buffer size.
+	 */
+	if (sec_size == 0)
```
This is needed because some keys will have leading zeros we must preserve. There is a comment few lines below. Let me know if you want me to move it

clementfaure replied:
I'm gonna check it

clementfaure replied:
We cannot rely on the bignum size as it could be smaller than the security key size. In this case, we need to specify the security key to properly do the bignum to bin translation and eventually pad the result. It was already done before :
https://github.com/OP-TEE/optee_os/blob/8e9d8acca4f8141b72d2d32846317954124c17ec/core/drivers/crypto/caam/acipher/caam_ecc.c#L108-L110

jenswi-linaro replied:
Please leave an explanation like this in the comment above for future reference.

etienne-lms commented on core/drivers/crypto/caam/include/caam_desc_helper.h:
```diff
@@ -234,8 +234,9 @@ static inline void dump_desc(uint32_t *desc)
-	(CMD_FIFO_STORE_TYPE | FIFO_STORE_OUTPUT(src) | FIFO_STORE_LENGTH(len))
+#define FIFO_ST(cla, src, len)                                                 \
+	(CMD_FIFO_STORE_TYPE | CMD_CLASS(cla) | FIFO_STORE_OUTPUT(src) |       \
+	FIFO_STORE_LENGTH(len))
```
mind indentation (for consistency with other macro definitions)

clementfaure replied:
Fixed

etienne-lms reviewed (commented):
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits 
"drivers: caam: add class field to FIFO_ST macro" (with comment addressed),
"drivers: caam: add missing header" (add `Fixes: 38923d487567 ("drivers: caam: implement CAAM DMA Object")` in commit message),
"drivers: caam: fix DSA_DUMPDESC macro" (add `Fixes: 4b383f736e9e ("drivers: caam: implement NXP CAAM Driver - DSA"` is commit message)

`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"core: move CFG_CORE_BIGNUM_MAX_BITS default definition".

clementfaure commented on core/drivers/crypto/caam/include/caam_key.h:
```diff
@@ -0,0 +1,155 @@
+ * @data	Buffer input
+ * @size	Buffer input size
+ * @key		CAAM key structure to populate
+ * @size_sec	Security key size to deserialize, optional. If not needed,
```
Fixed

clementfaure commented:
Force-push: rebase
Push: fixup commits:
- @jenswi-linaro I tried to follow your recommandations. I went with `switch` instead of `if else` and remove duplicated testing.
- @jforissier I still get checkpatch warnings about macros `Macro argument reuse 'alg' - possible side-effects?` Should I go with GNU extensions?

clementfaure commented:
Force-push : checkpatch warnings fixed and macros reworked

clementfaure commented:
Any comments?

jenswi-linaro reviewed (commented):
One comment for now. Please squash in the fixups, it's too long ago for me to remember what I looked at so I'll need to go through all the (unreviewed) patches again anyway.

jenswi-linaro commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,753 @@
+		output_buffer_size = out_key->buf.length;
+	} else {
+		/*
+		 * For an un-blob operation, the input key size is the original
```
s/un-blob/non-blob/

clementfaure replied:
Ok

jenswi-linaro commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,753 @@
+	 */
+	status = caam_key_serialized_size(&blob, &serialized_size);
+	if (status)
+		goto out;
```
Nothing is allocated yet, why not just return the error?

clementfaure replied:
Let's return the error indeed

jenswi-linaro commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,753 @@
+		KEY_TRACE("Destination buffer is too short %zu < %zu", size,
+			  serialized_size);
+		status = CAAM_OUT_MEMORY;
+		goto out;
```
Nothing is allocated yet, why not just return the error?

clementfaure replied:
Let's return the error indeed

jenswi-linaro commented on core/drivers/crypto/caam/caam_key.c:
```diff
@@ -0,0 +1,753 @@
+	status = caam_key_operation_blob(key, &blob);
+	if (status) {
+		KEY_TRACE("Blob operation fail");
+		goto out;
```
Is freeing needed? Else just return the error.

clementfaure replied:
same

jenswi-linaro reviewed (commented):
With the comments addressed, for the commits:
- "drivers: caam: add CAAM key driver"
- "drivers: caam: add CAAM key support for RSA"
- "drivers: caam: add CAAM key support for DSA"
- "drivers: caam: add CAAM key support for ECC"
- "drivers: caam: add CAAM key support for DH"
please apply:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

clementfaure commented:
Force-push: rebase + fix compilation warning.

jforissier commented:
@clementfaure thank you, this is ready for merging now!
